cartographer_ros/cartographer_ros/assets_writer_main.cc

```diff
@@ -149,69 +151,82 @@ void Run(const string& pose_graph_filename, const string& bag_filename,
-      << "Trajectory does not contain any nodes.";
+  CHECK_EQ(pose_graph_proto.trajectory_size(), bag_filenames.size())
+      << "Pose graphs contains " << pose_graph_proto.trajectory_size()
+      << " trajectories while " << bag_filenames.size()
+      << " bags where provided. This tool requires one bag for each "
+         "trajectory in the same order as the correponding trajectories in the "
+         "pose graph proto.";
 
   const auto file_writer_factory = [](const string& filename) {
     return carto::common::make_unique<carto::io::StreamFileWriter>(filename);
   };
 
   carto::io::PointsProcessorPipelineBuilder builder;
-  carto::io::RegisterBuiltInPointsProcessors(trajectory_proto,
+  // TODO(hrapp): Pass all trajectories to the PointsProcessors.
+  carto::io::RegisterBuiltInPointsProcessors(pose_graph_proto.trajectory(0),
                                              file_writer_factory, &builder);
   std::vector<std::unique_ptr<carto::io::PointsProcessor>> pipeline =
       builder.CreatePipeline(
           lua_parameter_dictionary.GetDictionary("pipeline").get());
 
-  tf2_ros::Buffer tf_buffer(::ros::DURATION_MAX);
-
-  if (FLAGS_use_bag_transforms) {
-    LOG(INFO) << "Pre-loading transforms from bag...";
-    ReadTransformsFromBag(bag_filename, &tf_buffer);
-  }
-
-  if (!urdf_filename.empty()) {
-    ReadStaticTransformsFromUrdf(urdf_filename, &tf_buffer);
-  }
-
   const string tracking_frame =
       lua_parameter_dictionary.GetString("tracking_frame");
-  const auto transform_interpolation_buffer =
-      carto::transform::TransformInterpolationBuffer::FromTrajectory(
-          trajectory_proto);
-  LOG(INFO) << "Processing pipeline...";
   do {
-    rosbag::Bag bag;
-    bag.open(bag_filename, rosbag::bagmode::Read);
-    rosbag::View view(bag);
-    const ::ros::Time begin_time = view.getBeginTime();
-    const double duration_in_seconds = (view.getEndTime() - begin_time).toSec();
-
-    for (const rosbag::MessageInstance& message : view) {
-      if (message.isType<sensor_msgs::PointCloud2>()) {
-        HandleMessage(*message.instantiate<sensor_msgs::PointCloud2>(),
-                      tracking_frame, tf_buffer,
-                      *transform_interpolation_buffer, pipeline);
+    for (size_t trajectory_id = 0; trajectory_id < bag_filenames.size();
+         ++trajectory_id) {
+      const carto::mapping::proto::Trajectory& trajectory_proto =
+          pose_graph_proto.trajectory(trajectory_id);
+      const string& bag_filename = bag_filenames[trajectory_id];
+      LOG(INFO) << "Processing " << bag_filename << "...";
+      if (trajectory_proto.node_size() == 0) {
+        continue;
       }
-      if (message.isType<sensor_msgs::MultiEchoLaserScan>()) {
-        HandleMessage(*message.instantiate<sensor_msgs::MultiEchoLaserScan>(),
-                      tracking_frame, tf_buffer,
-                      *transform_interpolation_buffer, pipeline);
+      tf2_ros::Buffer tf_buffer(::ros::DURATION_MAX);
```

Comment:
Should there be a `TODO` that this is known to be very inefficient and should be fixed?